### PR TITLE
Deal with trailing slash in tempRoot

### LINKED ISSUE #19191: tempRoot flag with trailing slash causes errors.
Running an IT with a tempRoot flag which ends in a '/' in Java will cause the pipeline to fail with a path validation error of some sort. Make tempRoot tolerant to trailing slashes.

Imported from Jira [BEAM-5721](https://issues.apache.org/jira/browse/BEAM-5721). Original Jira may contain additional context.
Reported by: jaku.

### PR DESCRIPTION
**Pull request by Abacn**
Fixes #19191

* Add FileSystems.matchNewDirectory and replace the raw string concatenate throughout the code base

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
Run Java PreCommit

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 1]
```diff
@@ -316,6 +317,24 @@ public void testInvalidSchemaMatchNewResource() {
     assertEquals("file", FileSystems.matchNewResource("c:/tmp/f1", false));
   }
 
+  @Test
```
Comment (Abacn) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java:320]: This unit test fails on windows. Will investigate further

Comment (Abacn) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java:320]: error on windows:

```
org.junit.ComparisonFailure: expected:<[/abc/d/]> but was:<[E:\abc\d\]>
	at org.junit.Assert.assertEquals(Assert.java:117)
```

Fixed by comparing resourceId and not converting to String, which will give OS specific valid paths

**Comment by Abacn** [comment]
assign set of reviewers

**Comment by shunping** [comment]
ack.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java [hunk 2]
```diff
@@ -74,8 +74,13 @@ public class TestDataflowRunner extends PipelineRunner<DataflowPipelineJob> {
   public static TestDataflowRunner fromOptions(PipelineOptions options) {
     TestDataflowPipelineOptions dataflowOptions = options.as(TestDataflowPipelineOptions.class);
     String tempLocation =
-        Joiner.on("/")
-            .join(dataflowOptions.getTempRoot(), dataflowOptions.getJobName(), "output", "results");
+        FileSystems.matchNewDirectory(
+                dataflowOptions.getTempRoot(), dataflowOptions.getJobName(), "output", "results")
+            .toString();
+    // to keep exact same behavior prior to matchNewDirectory introduced
+    if (tempLocation.endsWith("/") || tempLocation.endsWith("\\")) {
```
Comment (shunping) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java:81]: Can we use `File.separator` here? 

Reference: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java#L216

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java [hunk 3]
```diff
@@ -316,6 +317,25 @@ public void testInvalidSchemaMatchNewResource() {
...
+  public void testMatchNewDirectory() {
```
Comment (shunping) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java:321]: Does this test only work in Linux/MacOS? Do we need another test for Windows?

Comment (Abacn) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileSystemsTest.java:321]: this test works for all platforms (we have GHA for it). The initial commit was failing on windows but then fixed in followup.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java (hunk 2 above)
Comment (Abacn) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java]: change applied

**Comment by Abacn** [comment]
java precommit flaky test failed: `RemoteExecutionTest. testMetrics` ;
IOs PreCommit flaky test `CallTest.givenTeardownTimeout_throwsError` opened #29724. Both not related to this change.

**Pull request merged**